File: common/src/main/java/org/valkyrienskies/clockwork/content/forces/ReactionWheelController.java
package org.valkyrienskies.clockwork.content.forces;

import it.unimi.dsi.fastutil.Pair;
import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
import org.jetbrains.annotations.NotNull;
import org.joml.Matrix3dc;
import org.joml.Vector3d;
import org.joml.Vector3dc;
import org.valkyrienskies.clockwork.content.contraptions.reaction_wheel.ReactionWheelCreateData;
import org.valkyrienskies.clockwork.content.contraptions.reaction_wheel.ReactionWheelData;
import org.valkyrienskies.clockwork.content.contraptions.reaction_wheel.ReactionWheelUpdateData;
import org.valkyrienskies.core.api.ships.PhysShip;
import org.valkyrienskies.core.api.ships.ServerShip;
import org.valkyrienskies.core.api.ships.properties.ShipTransform;
import org.valkyrienskies.core.impl.api.ShipForcesInducer;
import org.valkyrienskies.core.impl.game.ships.PhysShipImpl;

import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentLinkedQueue;

public class ReactionWheelController implements ShipForcesInducer {

    final Int2ObjectOpenHashMap<ReactionWheelData> reactionwheelData = new Int2ObjectOpenHashMap<>();

    private final ConcurrentHashMap<Integer, ReactionWheelUpdateData> reactionwheelUpdateData = new ConcurrentHashMap<>();

    private final ConcurrentLinkedQueue<Pair<Integer, ReactionWheelCreateData>> createdRWs = new ConcurrentLinkedQueue<>();

    private final ConcurrentLinkedQueue<Integer> removedRWs = new ConcurrentLinkedQueue<>();

    private int nextRWID = 0;

    public static ReactionWheelController getOrCreate(ServerShip ship) {
        if (ship.getAttachment(ReactionWheelController.class) == null) {
            ship.saveAttachment(ReactionWheelController.class, new ReactionWheelController());
        }
        return ship.getAttachment(ReactionWheelController.class);
    }

    @Override
    public void applyForces(@NotNull PhysShip physShip) {
        while (!createdRWs.isEmpty()) {
            final Pair<Integer, ReactionWheelCreateData> createData = createdRWs.remove();
            reactionwheelData.put(createData.left(), new ReactionWheelData(
                createData.right().wheelPos(),
                createData.right().wheelAxis(),
                createData.right().wheelSpeed(),
                createData.right().spinup(),
                createData.right().spindown(),
                createData.right().active(),
                createData.right().sourceSpeed()
            ));
        }
        while (!removedRWs.isEmpty()) {
            reactionwheelData.remove((int) removedRWs.remove());
        }

        reactionwheelUpdateData.forEach((id, data) -> {
            ReactionWheelData physData = reactionwheelData.get(id);
            if (physData == null) {
                return;
            }
            physData.wheelSpeed = data.speed();
            physData.sourceSpeed = data.sourceSpeed();
        });

        reactionwheelUpdateData.clear();

        for (ReactionWheelData physData : reactionwheelData.values()) {
//            if (physData.wheelAxis.x() == 1) {
//                physData.active = Math.abs((((PhysShipImpl) physShip).getPoseVel().getOmega().x())) >= 10;
//            } else if (physData.wheelAxis.y() == 1) {
//                physData.active = Math.abs((((PhysShipImpl) physShip).getPoseVel().getOmega().y())) >= 10;
//            } else if (physData.wheelAxis.z() == 1) {
//                physData.active = Math.abs((((PhysShipImpl) physShip).getPoseVel().getOmega().z())) >= 10;
//            }
            //FOR TESTING
            physData.active = true;


            if (physData.sourceSpeed != 0) {
                if (physData.active) {

                    Vector3dc torque = computeTorque(physShip.getTransform(), physData, ((PhysShipImpl) physShip).getPoseVel().getOmega(), ((PhysShipImpl) physShip));
//                    Vector3dc troque = computeResistance(((PhysShipImpl) physShip), physData);
                    if (torque.isFinite()) {
                        physShip.applyInvariantTorque(torque);
                    }
                }
            }
        }
    }

    private Vector3dc computeTorque(ShipTransform physTransform, ReactionWheelData physWheel, Vector3dc omega, PhysShipImpl physShip) {
        Vector3dc prevAngMomentumRelWheel = physWheel.prevAngMomentum;
        Vector3dc wheelPos = physWheel.wheelPos;

        Vector3dc wheelAxis = new Vector3d(physWheel.wheelAxis);
        double wheelSpeed = physWheel.wheelSpeed;


        double wheelMass = 18000;

        // 1/2 * Mass * (Outer Wheel Radius^2 + Total Wheel Radius^2)
        double wheelInertia = (0.5 * wheelMass) * (Math.pow(0.25, 2) + Math.pow(0.75, 2));

        double rotVel = wheelSpeed * ((2 * Math.PI) / 60);

        Vector3dc r = new Vector3d(wheelPos).sub(physTransform.getPositionInShip()).rotate(physTransform.getShipToWorldRotation());
        Vector3dc angularMomentumRelWheel = new Vector3d(wheelAxis).mul(rotVel).mul(wheelInertia);

        // Add to convert from momentum relative to wheel into relative to ship
        Vector3dc momentumModifier = new Vector3d(omega).cross(r).mul(wheelMass);

        Vector3dc angularMomentumRelShip = new Vector3d(angularMomentumRelWheel).add(momentumModifier);
        Vector3dc prevAngularMomentumRelShip = new Vector3d(prevAngMomentumRelWheel).add(momentumModifier);

        Vector3dc torque = new Vector3d(prevAngularMomentumRelShip).sub(angularMomentumRelShip).div(1 / 60.0);

        physWheel.prevAngMomentum = angularMomentumRelWheel;

        return torque;
    }

    public int addReactionWheel(ReactionWheelCreateData data) {
        int id = nextRWID++;
        createdRWs.add(Pair.of(id, data));
        return id;
    }

    public void removeReactionWheel(int id) {
        removedRWs.add(id);
    }

    public void updateReactionWheel(int id, ReactionWheelUpdateData data) {
        reactionwheelUpdateData.put(id, data);
    }
}

File: common/src/main/kotlin/org/valkyrienskies/clockwork/content/forces/ReactionWheelController.kt
package org.valkyrienskies.clockwork.content.forces

import com.fasterxml.jackson.annotation.JsonAutoDetect
import com.fasterxml.jackson.annotation.JsonIgnore
import org.joml.Vector3d
import org.joml.Vector3dc
import org.valkyrienskies.clockwork.ClockworkConfig
import org.valkyrienskies.clockwork.content.physicalities.reactionwheel.data.ReactionWheelCreateData
import org.valkyrienskies.clockwork.content.physicalities.reactionwheel.data.ReactionWheelData
import org.valkyrienskies.clockwork.content.physicalities.reactionwheel.data.ReactionWheelUpdateData
import org.valkyrienskies.core.api.ships.LoadedServerShip
import org.valkyrienskies.core.api.ships.PhysShip
import org.valkyrienskies.core.api.world.PhysLevel
import org.valkyrienskies.core.impl.game.ships.PhysShipImpl
import java.util.concurrent.ConcurrentLinkedQueue

@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
class ReactionWheelController(
    override val appliers: HashMap<Int, ReactionWheelData> = HashMap(),
    override val applierUpdateData: ConcurrentLinkedQueue<Pair<Int, ReactionWheelUpdateData>> = ConcurrentLinkedQueue(),
    override val createdAppliers: ConcurrentLinkedQueue<Pair<Int, ReactionWheelCreateData>> = ConcurrentLinkedQueue(),
    override val removedAppliers: ConcurrentLinkedQueue<Int> = ConcurrentLinkedQueue(),
    override var nextApplierID: Int = 0
) : MultiInstanceForceApplier<ReactionWheelUpdateData, ReactionWheelData, ReactionWheelCreateData> {

    @JsonIgnore
    private var previousTickShipOmega: Vector3dc = Vector3d()
    @JsonIgnore
    val pendingMomentumConsumptionQueue = HashMap<Int, ConcurrentLinkedQueue<Double>>()

    override fun physTick(physShip: PhysShip, physLevel: PhysLevel) {
        super.physTick(physShip, physLevel)

        for (wheelID in appliers.keys) {
            val (torque, torque2) = conserveMomentum(physShip as PhysShipImpl, appliers[wheelID]!!)
            if (torque.isFinite && !torque.length().isNaN()) {
                physShip.applyWorldTorque(torque)
                physShip.applyWorldTorque(torque2)
            }
        }

        previousTickShipOmega = physShip.omega
    }

    fun conserveMomentum(physShip: PhysShipImpl, wheel: ReactionWheelData): Pair<Vector3dc, Vector3dc> {
        val wheelPos: Vector3dc = Vector3d(wheel.position!!)
        val wheelDir: Vector3dc = Vector3d(wheel.direction!!)
        val wheelAxisInWorld: Vector3dc = physShip.transform.shipToWorld.transformDirection(wheelDir, Vector3d()).normalize()

        val wheelSpeed = wheel.currentRPM / 60.0 * 2.0 * Math.PI
        val wheelPrevious = wheel.previousRPM / 60.0 * 2.0 * Math.PI

        val wheelMass = ClockworkConfig.SERVER.reactionWheelEffectiveness * 1_920_000.0.coerceAtMost(9600 * physShip.mass)
        // 1/2 * m * (r_2^2 + r_1^2)
        val wheelInertia = (0.5 * wheelMass) * (Math.pow(0.25, 2.0) + Math.pow(0.75, 2.0))
        // 1/12 * m * (3 * (r_2^2 + r_1^2) + h^2)
        val offAxisInertia = (1/12.0 * wheelMass) * (3*(Math.pow(0.25, 2.0) + Math.pow(0.75, 2.0)) + Math.pow(0.5, 2.0))

        val newW: Vector3dc = wheelAxisInWorld.mul(wheelSpeed, Vector3d())
        val deltaW: Vector3dc = newW.normalize(Vector3d()).cross(physShip.angularVelocity.div( 60.0, Vector3d()), Vector3d())

        val axialTorque = wheelAxisInWorld.mul((wheelSpeed - wheelPrevious) * wheelInertia / 20.0, Vector3d())

        val torqueInduced: Vector3dc = deltaW.mul(offAxisInertia, Vector3d())

        return torqueInduced to axialTorque
    }

    companion object {
        fun getOrCreate(ship: LoadedServerShip): ReactionWheelController? {
            if (ship.getAttachment(ReactionWheelController::class.java) == null) {
                ship.setAttachment(ReactionWheelController())
            }
            return ship.getAttachment(ReactionWheelController::class.java)
        }
    }
}
